torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py

```diff
@@ -0,0 +1,408 @@
+            )
+        if func is torch.ops.aten.clone.default:
+            return return_and_correct_aliasing(
+                func, args, kwargs, args[0]._apply_fn_to_data(torch.clone)
+            )
+
+        raise NotImplementedError(
+            f"IntxWeightAQTLayout dispatch: attempting to run {func}, this is not supported"
+        )
+
+    def __tensor_flatten__(self):
+        if _target_equal(self.get_layout_type().target, "native"):
+            return ["packed_weight"], [self.get_layout_type()]
+
+        # fallback
+        assert _target_equal(self.get_layout_type().target, "fallback")
+        if self.zero_point is None:
+            return ["packed_weight", "scale"], [self.get_layout_type()]
+        return ["packed_weight", "scale", "zero"], [self.get_layout_type()]
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        packed_weight, scale, zero_point = (
+            tensor_data_dict["packed_weight"],
+            tensor_data_dict.get("scale", None),
+            tensor_data_dict.get("zero_point", None),
+        )
+        (layout_type,) = tensor_attributes
+        return cls(packed_weight, scale, zero_point, layout_type)
+
+
+def _linear_int8_dynamic_activation_intx_weight_check(
+    input_tensor, weight_tensor, bias
+):
+    layout_type = weight_tensor.layout_tensor.layout_type
+    return isinstance(layout_type, IntxWeightLayoutType) and bias is None
+
+
+def _linear_int8_dynamic_activation_intx_weight_fallback_impl(
+    input_tensor, weight_tensor, bias
+):
+    assert weight_tensor.layout_tensor.layout_type.target == "fallback"
+    assert bias is None
+
+    def _impl_2d(input_tensor, weight_tensor):
+        assert input_tensor.dim() == 2
+        assert weight_tensor.dim() == 2
+
+        weight_qvals = weight_tensor.layout_tensor.packed_weight.to(torch.int32)
+        weight_scales = weight_tensor.layout_tensor.scale
+        weight_zeros = weight_tensor.layout_tensor.zero_point
+        group_size = weight_tensor.layout_tensor.layout_type.group_size
+        has_weight_zeros = weight_zeros is not None
+        m, k = input_tensor.shape
+        n, k_ = weight_tensor.shape
+        assert k_ == k
+
+        weights_dequantized = dequantize_per_channel_group(
+            w_int8=weight_qvals,
+            scales=weight_scales,
+            zero_points=(
+                weight_zeros if has_weight_zeros else torch.zeros_like(weight_scales)
+            ),
+            quant_min=None,  # TODO: why is this an arg for this function
+            quant_max=None,  # TODO: why is this an arg for this function
+            dtype=None,  # TODO: why is this an arg for this function
+            group_size=group_size,
+            output_dtype=torch.float32,
+        )
+
+        # Quantize activations
+        activation_scales, activation_zeros = choose_qparams_affine(
```

Comment:
dynamic quantization should be reusing affine quantized tensor, example: https://github.com/pytorch/ao/blob/900f9acccb8bc3c102d3a7095d1b1f0355057bec/torchao/quantization/quant_api.py#L586

why is this calling these functions here?

Reply:
That function doesn't look equivalent?  It looks like the quantization is symmetric.

Reply:
you can choose a different mapping type,
I mean we can use https://github.com/pytorch/ao/blob/900f9acccb8bc3c102d3a7095d1b1f0355057bec/torchao/quantization/quant_api.py#L613 to do activation quantization, maybe "Dynamic Activation Quantization + Weight Quantization" in https://github.com/pytorch/ao/issues/391 is clearer in explaining this, please let me know if it makes sense

Reply:
Sorry, I don't follow what you're suggesting.  I have weights_dequantized on line 260, and I need activations_dequantized so I can call torch.matmul(activations_dequantized, weights_dequantized.transpose(1, 0)) on line 296.

I'm not sure what I'm suppose to replace the code that generates activations_dequantized with.